### ISSUE TITLE
Fission fn return partial json incase the json reponse is greater than 5000 lines - Version 1.5

### ISSUE THREAD

**Issue by hd40910** [fission/fission#1409]
Error:

Nov 15 17:21:23 test dockerd-current[33402]: 15-11-2019 15:21:23[INFO ][sdfd:None]: generated response headers:
Nov 15 17:21:23 test dockerd-current[33402]: [39B blob data]
Nov 15 17:21:23 test dockerd-current[33402]: [23B blob data]
Nov 15 17:21:23 test dockerd-current[33402]: [69B blob data]
Nov 15 17:21:23 test dockerd-current[33402]: [1B blob data]
Nov 15 17:21:23 test dockerd-current[33402]:
Nov 15 17:21:23 test dockerd-current[33402]: Client 6 hit errno 104
Nov 15 17:21:23 test dockerd-current[33402]: 2019/11/15 15:21:23 [error] **29913#29913: *2317320 upstream prematurely closed connection while sending to client, client:** 10.75.217.28, server: _, request: "POST /sdf/sdfds/sdfds HTTP/1.1", upstream: "http://10.244.0.37:8888/sdf/sdf/sdfds", host: "test:30001"
Nov 15 17:21:23 test dockerd-current[33402]: asdas - [asdsad] - - [15/Nov/2019:15:21:23 +0000] "POST /asd/asd/asd HTTP/1.1" 200 954 "-" "PostmanRuntime/7.19.0" 760228 0.949 [fission-router-80] 10.244.0.37:8888 3924 0.084 200 86a8c7b659714bc3c45981abc35c629a
Nov 15 17:21:23 test kubelet[31171]: W1115 17:21:23.944912   31171 kubelet_getters.go:264] Path "/var/lib/kubelet/pods/df4ccf64-06ce-11ea-bfa4-005056aff425/volumes" does not exist
Nov 15 17:21:23 test kubelet[31171]: W1115 17:21:23.944942   31171 kubelet_getters.go:264] Path "/var/lib/kubelet/pods/df4ccf64-06ce-11ea-bfa4-005056aff425/volumes" does not exist
Nov 15 17:21:23 test kubelet[31171]: E1115 17:21:23.945117   31171 kubelet_volumes.go:133] Failed to remove orphaned pod "df4ccf64-06ce-11ea-bfa4-005056aff425" dir; err: remove /var/lib/kubelet/pods/df4ccf64-06ce-11ea-bfa4-005056aff425/volume-subpaths/config-volume/controller/0: device or resource busy
Nov 15 17:21:23 test kubelet[31171]: W1115 17:21:23.945193   31171 kubelet_getters.go:264] Path "/var/lib/kubelet/pods/ec9550d1-06ad-11ea-98d7-005056aff425/volumes" does not exist
Nov 15 17:21:23 test kubelet[31171]: W1115 17:21:23.945202   31171 kubelet_getters.go:264] Path "/var/lib/kubelet/pods/ec9550d1-06ad-11ea-98d7-005056aff425/volumes" does not exist

Input :
![image](https://user-images.githubusercontent.com/17830882/68954863-daa1cd00-07ea-11ea-9423-7a259bb1e9dc.png)

Output : 
![image](https://user-images.githubusercontent.com/17830882/68954829-cb228400-07ea-11ea-9aab-b92a553c398d.png)

**Comment by hd40910** [comment]
@viveksinghggits  Can you please help check ?

**Comment by viveksinghggits** [comment]
Hi @hd40910 
can you please mentioned the when are you getting these errors and how can we reproduce them. Assume that noone knows about the context you are talking about.

**Comment by hd40910** [comment]
Hi @viveksinghggits 

One out of 3 http trigger calls to a fission function returns partial data.
Given above is an example where we have a function which simply accepts a 5000 line valid json file as input.But as attached only partial data is getting returned.

The only error seen is "*29913#29913: 2317320 upstream prematurely closed connection while sending to client, client: 10.75.217.28, server: in" the k8 logs

This is happening once in few calls.

**Comment by viveksinghggits** [comment]
Let me re-create this I will update soon.

**Comment by life1347** [comment]
@hd40910 could you also try with 1.6.0 since 1.5.0 has a known issue in the executor.

**Comment by hd40910** [comment]
@life1347 This gets recreated in 1.6.0. Tested

![image](https://user-images.githubusercontent.com/17830882/69008801-96dacf00-0974-11ea-890a-168b0ffdc829.png)

**Comment by msshroff** [comment]
Hi @life1347  

Issue still persists post using fission client version 1.6.0
<pre>
# fission -v
client:
  fission/core:
    BuildDate: "2019-10-10T18:24:57Z"
    GitCommit: 08b6d43ac1f1fa72854438aa6db636fd04c93098
    Version: 1.6.0
server:
  fission/core:
    BuildDate: "2019-10-10T18:24:57Z"
    GitCommit: 08b6d43ac1f1fa72854438aa6db636fd04c93098
    Version: 1.6.0
</pre>


Specs for the fission function is below: 
<pre>
# cat function-atomiq-span.yaml
include:
- atomiq_span.py
kind: ArchiveUploadSpec
name: atomiq-span-py-ltbo

---
apiVersion: fission.io/v1
kind: Package
metadata:
  creationTimestamp: null
  name: atomiq-span-py-oopi
  namespace: default
spec:
  deployment:
    checksum: {}
    type: url
    url: archive://atomiq-span-py-ltbo
  environment:
    name: atomiq-py-ssh
    namespace: fission-function
  source:
    checksum: {}
status:
  buildstatus: none

---
apiVersion: fission.io/v1
kind: Function
metadata:
  creationTimestamp: null
  name: atomiq-span
  namespace: default
spec:
  InvokeStrategy:
    ExecutionStrategy:
      ExecutorType: newdeploy
      MaxScale: 3
      MinScale: 1
      SpecializationTimeout: 360
      TargetCPUPercent: 80
    StrategyType: execution
  configmaps: null
  environment:
    name: atomiq-py-ssh
    namespace: fission-function
  package:
    packageref:
      name: atomiq-span-py-oopi
      namespace: default
  resources: {}
  secrets: null
</pre>

Logs in ingress when there is **NO** error
<pre>
2019/11/18 07:01:08 [warn] 730#730: *128552 a client request body is buffered to a temporary file /tmp/client-body/0000000087, client: 10.244.0.1, server: _, request: "POST /default/atomiq_py_ssh/atomiq_span HTTP/1.1", host: "ilgss0279:30001"
10.244.0.1 - [10.244.0.1] - - [18/Nov/2019:07:01:08 +0000] "POST /default/atomiq_py_ssh/atomiq_span HTTP/1.1" 200 190418 "-" "curl/7.29.0" 222377 0.022 [fission-router-80] 10.244.0.82:8888 190393 0.020 200 15723581e1b12a7ef98aece4a1829b02
</pre>



Logs in ingress when there **IS** error
<pre>
2019/11/18 07:01:43 [warn] 726#726: *129395 a client request body is buffered to a temporary file /tmp/client-body/0000000091, client: 10.244.0.1, server: _, request: "POST /default/atomiq_py_ssh/atomiq_span HTTP/1.1", host: "ilgss0279:30001"
2019/11/18 07:01:43 [error] 726#726: *129395 upstream prematurely closed connection while reading upstream, client: 10.244.0.1, server: _, request: "POST /default/atomiq_py_ssh/atomiq_span HTTP/1.1", upstream: "http://10.244.0.82:8888/default/atomiq_py_ssh/atomiq_span", host: "ilgss0279:30001"
10.244.0.1 - [10.244.0.1] - - [18/Nov/2019:07:01:43 +0000] "POST /default/atomiq_py_ssh/atomiq_span HTTP/1.1" 200 3949 "-" "curl/7.29.0" 222377 0.059 [fission-router-80] 10.244.0.82:8888 3924 0.058 200 ec04d1309f241e82fbf1b77f43919a41
</pre>

**Comment by life1347** [comment]
@msshroff do you know what's the total size of return JSON? I' m wondering if you're hitting the body size limit of Nginx. Its default size is 1m.

https://github.com/kubernetes/ingress-nginx/blob/master/docs/user-guide/nginx-configuration/annotations.md#custom-max-body-size

**Comment by msshroff** [comment]
Nginx limitation does not seem to be the issue.
This is an intermittent issue. Problems happens for roughly 1 in 5 requests. Works well for rest 4 with the same sized request body.

**Comment by life1347** [comment]
@msshroff thanks for your response. I'm trying to repro the problem and hope I can find something suspicious. At the time, could you do me a favor to test if you send requests directly to Fission router, does the same situation happen? In this way, we can narrow down the range of the problem. Thanks.

**Comment by msshroff** [comment]
hi @life1347 

Yes, same is reproduced by hitting router directly bypassing the ingress ( nginx)

**Comment by msshroff** [comment]
Also just checked the size of the workload for which its failing. The test which i ran is having 228K. So does not seem to be related to 1M default size limitation of nginx

**Comment by life1347** [comment]
@msshroff Thanks for the update. I can reproduce the problem.

**Comment by life1347** [comment]
After some experiments, I've found the root cause of the problem. 

Here are the things I've done to troubleshoot this problem.

1. Send requests to function pod directly and thanks to @msshroff help, we can narrow down the problem to the router.
2. Try to read the full response in RoundTripper and notice an error occurred when reading resp body.

Here is the code snippet I used to read resp body.

```
body, err := ioutil.ReadAll(resp.Body)
roundTripper.logger.Info("response body", zap.String("body", string(body)),
    zap.Any("header", resp.Header), zap.Error(err), zap.Int("context-timeout", roundTripper.timeout))
```

And received `context canceled` error while we still have enough time to before context timeout.

```
"error": "context canceled", "context-timeout": 60
```

The situation is clear now, the connection was closed due to the context was canceled while reading the response.

https://github.com/fission/fission/blob/d8b386a812971e03717e8def32cc3fd77dbfeb5c/pkg/router/functionHandler.go#L284

Comment out `closeCtx()`, and function works as expected.

<img width="1534" alt="Screenshot 2019-11-19 17 20 15" src="https://user-images.githubusercontent.com/202578/69133471-e9d99100-0af0-11ea-92a3-bbbb2c0127e8.png">

However, here is a remaining question I haven't figure out and need more investigation: Why some requests success?
